lib/logic.py

```diff
@@ -0,0 +1,152 @@
+import copy
+from dataclasses import dataclass
+import typing
+
+
+#
+# Traversal helpers and documentation
+#
+def get_node_children(node) -> list:
+    return node[":children"] if ":children" in node else []
+
+
+def build_basic_node_type_checker(step: str):
+    def is_node_of_type(node) -> bool:
+        return node[":step"] == step
+    return is_node_of_type
+
+
+is_root_node = build_basic_node_type_checker(":root")
+
+is_asset_node = build_basic_node_type_checker(":asset")
+# ':ticker'
+# rest are ephemeral/derived from :ticker:
+# ':name'
+# ':has_marketcap'
+# ':exchange'
+# ':price'
+# ':dollar_volume'
+
+is_if_node = build_basic_node_type_checker(":if")
+# All :children are definitely if-child
+
+is_if_child_node = build_basic_node_type_checker(":if-child")
+# ':is-else-condition?'
+# if not an else block, see `is_conditional_node`
+
+is_equal_weight_node = build_basic_node_type_checker(":wt-cash-equal")
+is_specified_weight_node = build_basic_node_type_checker(":wt-cash-specified")
+
+
+def is_weight_node(node):
+    return is_equal_weight_node(node) or is_specified_weight_node(node)
+
+
+is_group_node = build_basic_node_type_checker(":group")
+# ':name'
+# ':collapsed?'
+
+is_filter_node = build_basic_node_type_checker(":filter")
+# ':select?': should always be True... not sure why this is here.
+# ':select-fn': :bottom or :top
+# ':select-n': str, annoyingly. Always an int
+
+# ':sort-by?': should always be True... not sure why this is here.
+# ':sort-by-fn'
+# ':sort-by-window-days': str, annoyingly
+
+
+def is_conditional_node(node) -> bool:
+    """
+    if-child and not else
+
+    # ':lhs-fn'
+    # ':lhs-window-days'
+    # ':lhs-val'
+
+    # ':comparator'
+
+    # ':rhs-fn'
+    # ':rhs-fixed-value?'
+    # ':rhs-window-days'
+    # ':rhs-val'
+    """
+    return is_if_child_node(node) and not node[":is-else-condition?"]
+
+
+def get_lhs_ticker(node) -> typing.Optional[str]:
+    return node[':lhs-val']
+
+
+def get_rhs_ticker(node) -> typing.Optional[str]:
+    # yes, rhs behaves differently than lhs.
+    if not node.get(':rhs-fixed-value?', type(node[':rhs-val']) != str):
+        return node[":rhs-val"]
+
+
+class ComposerIndicatorFunction:
+    # TODO: fill in missing symbols
+    CURRENT_PRICE = ":current-price"
+    CUMULATIVE_RETURN = ":cumulative-return"
+    # STANDARD_DEVIATION_PRICE = ":todo"
+    # STANDARD_DEVIATION_RETURNS = ":todo"
+    # MAX_DRAWDOWN = ":todo"
+    MOVING_AVERAGE_PRICE = ":moving-average-price"
+    # MOVING_AVERAGE_RETURNS = ":todo"
+    # EMA_PRICE = ":todo"
+    RSI = ":relative-strength-index"
+
+
+class ComposerComparison:
+    LTE = ":lte"
+    LT = ":lt"
+    GTE = ":gte"
+    GT = ":gt"
+    EQ = ":eq"
+
+
+def get_ticker_of_asset_node(node) -> str:
+    return node[':ticker']
+
+
+#
+# Tree path reducer/state thinker
```

Comment:
This (and below) is an example of a refactor of duplicated logic. I wanted to group all my weights-calculating logic (a complex bit of thinking) in 1 place so I could easily reference in VectorBT and Human outputs, and no doubt QC too.